src/dbnode/storage/shard.go

```diff
@@ -1503,23 +1507,35 @@ func (s *dbShard) insertSeriesBatch(inserts []dbShardInsert) error {
 			// this method (insertSeriesBatch) via `entryRefCountIncremented` mechanism.
 			entry.OnIndexPrepare()
 
-			id := entry.Series.ID()
-			tags := entry.Series.Tags().Values()
-
-			var d doc.Document
-			d.ID = id.Bytes() // IDs from shard entries are always set NoFinalize
-			d.Fields = make(doc.Fields, 0, len(tags))
-			for _, tag := range tags {
-				d.Fields = append(d.Fields, doc.Field{
-					Name:  tag.Name.Bytes(),  // Tags from shard entries are always set NoFinalize
-					Value: tag.Value.Bytes(), // Tags from shard entries are always set NoFinalize
-				})
+			// Don't insert cold writes into the index insert queue.
+			if s.coldWritesEnabled && writeType == series.ColdWrite {
+				indexBlockStart := s.reverseIndex.BlockStartForWriteTime(pendingIndex.timestamp)
+				// NB(bodu): We finalize this entry w/o decrementing the entry ref count
+				// unless the entry is marked as having its ref count incremented.
+				// Entries in the shard insert queue are either of:
+				// - new entries
+				// - existing entries that we've taken a ref on (marked as entryRefCountIncremented)
+				entry.OnIndexFinalize(indexBlockStart)
```

Comment:
Nice looking good.

I'd like to get a metric for this case and deploy just to see it in action before we merge. 

Getting close!